src/validation/violation.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 Helsing GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{self, Display, Formatter};
+
+use miette::{
+    Diagnostic, LabeledSpan, MietteError, MietteSpanContents, Severity, SourceCode, SourceSpan,
+    SpanContents,
+};
+
+/// Severity level of violation.
+#[derive(Clone, Debug, Copy, PartialEq, Eq, PartialOrd, Ord)]
+#[allow(missing_docs)]
+pub enum Level {
+    Info,
+    Warning,
+    Error,
+}
+
+/// Location of violation.
+#[derive(Default, PartialEq, Clone, Eq, Debug)]
+pub struct Location {
+    /// File that contains violation
+    pub file: Option<String>,
+    /// Package name of file containing violation
+    pub package: Option<String>,
+    /// Entity name containing the violation
+    pub entity: Option<String>,
+}
+
+/// Violation message.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Message {
+    /// Message describing violation
+    pub message: String,
+    /// Information on what went wrong
+    pub help: String,
+}
+
+impl std::error::Error for Message {}
+
+impl Display for Message {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", &self.message)
+    }
+}
+
+impl Diagnostic for Message {}
+
+/// Rule violation
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Violation {
+    /// Rule name that was violated
+    pub rule: String,
+    /// Level of violation
+    pub level: Level,
+    /// Message
+    pub message: Message,
+    /// Location where violation occured
+    pub location: Location,
+    /// Help text
+    pub info: String,
+}
+
+/// Alias for list of [`Violation`].
+pub type Violations = Vec<Violation>;
+
+impl std::error::Error for Violation {}
+
+impl Display for Violation {
```

Comment:
You should also be able to derive all of this with thiserror – as we do elsewhere in the crate